kernel/work.c

```diff
@@ -0,0 +1,966 @@
+			 * this loops back to wait on notify, or the thread
+			 * exits.
+			 *
+			 * We don't touch K_WORK_QUEUE_PLUGGABLE, so getting
+			 * here doesn't mean that the queue will allow new
+			 * submissions.
+			 */
+			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		}
+
+		if (work == NULL) {
+			/* Nothing's had a chance to add work since we took
+			 * the queue lock, and We didn't find work nor got
+			 * asked to stop.  Just go to sleep: when something
+			 * happens the work thread will be woken and we can
+			 * check again.
+			 */
+
+			(void)z_sched_wait(&queue->lock, key, &queue->notifyq,
+					   K_FOREVER, NULL);
+			continue;
+		}
+
+		k_spin_unlock(&queue->lock, key);
+
+		if (work != NULL) {
+			k_work_handler_t handler = work->handler;
+
+			__ASSERT_NO_MSG(handler != 0);
+
+			if (work_set_running(work, queue)) {
+				handler(work);
+				work_clear_running(work);
+			}
+
+			/* No longer referencing the work, so we can clear the
+			 * BUSY flag while we yield to prevent starving other
+			 * threads.
+			 */
+			key = k_spin_lock(&queue->lock);
+			atomic_clear_bit(&queue->flags, K_WORK_QUEUE_BUSY_BIT);
+			k_spin_unlock(&queue->lock, key);
+
+			k_yield();
+		}
+	}
+}
+
+void k_work_queue_start(struct k_work_q *queue,
+			 k_thread_stack_t *stack,
+			 size_t stack_size,
+			 int prio,
+			 const char *name)
+{
+	__ASSERT_NO_MSG(queue);
+	__ASSERT_NO_MSG(stack);
+	__ASSERT_NO_MSG(!atomic_test_bit(&queue->flags,
+					 K_WORK_QUEUE_STARTED_BIT));
+
+	sys_slist_init(&queue->pending);
+	z_waitq_init(&queue->notifyq);
+	z_waitq_init(&queue->drainq);
+
+	/* It hasn't actually been started yet, but all the state is in place
+	 * so we can submit things and once the thread gets control it's ready
+	 * to roll.
+	 */
+	atomic_set(&queue->flags, K_WORK_QUEUE_STARTED);
+
+	(void)k_thread_create(&queue->thread, stack, stack_size,
+			      work_queue_main, queue, NULL, NULL,
+			      prio, 0, K_NO_WAIT);
+
+	if (name) {
+		k_thread_name_set(&queue->thread, name);
+	}
+}
+
+int k_work_queue_drain(struct k_work_q *queue,
+		       bool plug)
+{
+	__ASSERT_NO_MSG(queue);
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	int ret = 0;
+	k_spinlock_key_t key = k_spin_lock(&queue->lock);
+
+	if (((atomic_get(&queue->flags)
+	      & (K_WORK_QUEUE_BUSY | K_WORK_QUEUE_DRAIN)) != 0U)
+	    || plug
+	    || !sys_slist_is_empty(&queue->pending)) {
+		atomic_set_bit(&queue->flags, K_WORK_QUEUE_DRAIN_BIT);
+		if (plug) {
+			atomic_set_bit(&queue->flags, K_WORK_QUEUE_PLUGGED_BIT);
+		}
+
+		notify_queue_locked(queue);
+		ret = z_sched_wait(&queue->lock, key, &queue->drainq,
+				   K_FOREVER, NULL);
+	} else {
+		k_spin_unlock(&queue->lock, key);
+	}
+
+	return ret;
+}
+
+int k_work_queue_unplug(struct k_work_q *queue)
+{
+	__ASSERT_NO_MSG(queue);
+
+	int ret = -EALREADY;
+	k_spinlock_key_t key = k_spin_lock(&queue->lock);
+
+	if (atomic_test_and_clear_bit(&queue->flags,
+				      K_WORK_QUEUE_PLUGGED_BIT)) {
```

Comment:
Yes, the flag manipulation in multiple places (`work_queue_main`, `queue_submit_locked`, `k_work_queue_drain` are under lock because they aren't atomic state manipulations.  `work_queue_drain` can break if `K_WORK_QUEUE_PLUGGED_BIT` changes while the lock is held.

We can get rid of the atomics entirely if spin_lock/unlock is a memory barrier in all cases.